src/feature-feedback.ts

```diff
@@ -0,0 +1,608 @@
+{
+  @property({ type: String }) featureIdentifier?: string;
+
+  @property({ type: String }) prompt = 'Do you find this feature useful?';
+
+  @property({ type: String }) buttonText = 'Beta';
+
+  @property({ type: Object }) recaptchaManager?: RecaptchaManagerInterface;
+
+  @property({ type: Object }) resizeObserver?: SharedResizeObserverInterface;
+
+  @property({ type: Object })
+  featureFeedbackService?: FeatureFeedbackServiceInterface;
+
+  @query('#beta-button') private betaButton!: HTMLButtonElement;
+
+  @query('#popup') private popup!: HTMLDivElement;
+
+  @state() private isOpen = false;
+
+  @state() private processing = false;
+
+  @state() private popupTopX = 0;
+
+  @state() private popupTopY = 0;
+
+  @state() private vote?: Vote;
+
+  @state() private voteSubmitted = false;
+
+  @state() private error?: TemplateResult;
+
+  @state() private voteNeedsChoosing = false;
+
+  @state() private recaptchaWidget?: RecaptchaWidgetInterface;
+
+  @query('#comments') private comments!: HTMLTextAreaElement;
+
+  private boundEscapeListener?: (this: Document, ev: KeyboardEvent) => any;
+
+  private boundScrollListener?: (this: Document, ev: Event) => any;
+
+  render() {
+    return html`
+      <button id="beta-button" @click=${this.showPopup} tabindex="0">
+        ${this.buttonText}
+        <span
+          class="beta-button-thumb upvote-button ${this.voteSubmitted
+            ? this.upvoteButtonClass
+            : ''}"
+          >${thumbsUp}</span
+        >
+        <span
+          class="beta-button-thumb downvote-button ${this.voteSubmitted
+            ? this.downvoteButtonClass
+            : ''}"
+          id="beta-button-thumb-down"
+          >${thumbsDown}</span
+        >
+      </button>
+      ${this.popupTemplate}
+    `;
+  }
+
+  firstUpdated(): void {
+    this.boundEscapeListener = this.handleEscape.bind(this);
+    this.boundScrollListener = this.handleScroll.bind(this);
+  }
+
+  updated(changed: PropertyValues): void {
+    if (changed.has('vote') && this.vote) {
+      this.error = undefined;
+      this.voteNeedsChoosing = false;
+    }
+    if (changed.has('resizeObserver')) {
+      const oldObserver = changed.get(
+        'resizeObserver'
+      ) as SharedResizeObserverInterface;
+      this.disconnectResizeObserver(oldObserver);
+    }
+  }
+
+  handleResize() {
+    if (!this.isOpen) return;
+    this.positionPopup();
+  }
+
+  handleScroll() {
+    if (!this.isOpen) return;
+    this.positionPopup();
+  }
+
+  disconnectedCallback(): void {
+    this.removeEscapeListener();
+    this.disconnectResizeObserver(this.resizeObserver);
+  }
+
+  private resizingElement = document.body;
+
+  private disconnectResizeObserver(observer?: SharedResizeObserverInterface) {
+    observer?.removeObserver({
+      handler: this,
+      target: this.resizingElement,
+    });
+  }
+
+  private setupScrollObserver() {
+    document.addEventListener('scroll', this.boundScrollListener!);
```

Comment:
q: what does the `!` at the end do?

Reply:
`this.boundScrollListener` is marked optional, but will effectively always have a value because we set it during `firstUpdated` so it's never really going to be undefined in practice. Using the `!` is just telling Typescript that that property will always be defined do you don't have to do `if (this.boundScrollListener) ...`.

Having said that, I think I can move the `!` up to the `boundScrollListener` definition so I don't need to do it here.